packages/app-builder-lib/src/util/MemoLazy.ts

```diff
@@ -0,0 +1,36 @@
+/**
+ * A reworked implementation of lazy-val (https://github.com/develar/lazy-val/blob/master/src/main.ts)
+ * that re-calculates the lazy `value` if the provided argument has changed.
+ */
+
+export class MemoLazy<S, V> {
+  private selected: S | undefined = undefined
+  private _value: Promise<V> | undefined = undefined
+
+  constructor(
+    private selector: () => S,
+    private creator: (selected: S) => Promise<V>
+  ) {}
+
+  get hasValue() {
+    return this._value !== undefined
+  }
+
+  get value(): Promise<V> {
+    const selected = this.selector()
+    if (this._value !== undefined && JSON.stringify(this.selected) === JSON.stringify(selected)) {
```

Comment:
This is an awesome class, thank you! Quick Q here, what if `this.selected` is not serializable into JSON?

It might be worth opening a PR for https://github.com/develar/lazy-val/blob/master/src/main.ts as well so that we could get this pushed upstream. I can ping develar to get it merged (he's usually pretty quick on those once I email him 🙂 )

Reply:
Yeah, I would have preferred to use either a shallow or deep equals, but I wanted to avoid adding too much new code for the PR.

I'll make this PR upstream and I'll switch to a shallow equals implementation over there! If develar is happy with it, I'll come back here to bump the package and use the newly-exported `MemoLazy`